### ISSUE TITLE
can not read from CAN

### ISSUE THREAD

**Issue by wjquigs11** [hatlabs/discussions#135]
I'm running the SH-ESP32-nmea2000-gateway code on a new SH-ESP32.
I cannot send or receive CAN bus data. My lab network has an ESP32 transmitting wind data, a CYD with external CAN transceiver, and an RPI running SK. All other devices are communicating normally.
Some of the things I have tried so far:
- running the same gateway code on CYD (obviously I changed defs for TX/RX GPIOs); everything works as expected
- swapping CAN H/L on the external connector
- sending data to the SH-ESP32 in Actisense format on UART works as expected
- swapping connectors
- swapping cables
- using jumpers directly to the board socket instead of using the connector
- connecting external CAN transceiver to available GPIOs works as expected
- I've tried using N2K bus 12V power instead of USB power; makes no difference

Am I missing something obvious? Is there something I have to do to enable the on-board CAN interface on GPIO 34/32?

**Comment by mairas** [comment]
I don't know what a CYD is, so can't reflect on that. But what is your lab NMEA 2000 setup like? How are you powering it?

Is the SH-ESP32 version 2.0.x or 2.1.0? The latter has CAN RX/TX indicator LEDs. If it's that one, what do the LEDs do?

Can you provide photos of the setup?

**Comment by wjquigs11** [comment]
I am using N2K t-connectors and terminators, powered by an AC mains->12VDC adapter. I have also tried powering the SH-ESP32 from USB alone. I measure 12VDC on the N2K power leads. I've also tried disconnecting 12V power from the network and running all the modules from USB.
The board is 2.1 and the only LED that's lit is power.
I'm using the same connectors for all of the modules so I'm fairly certain the wiring is correct.
![2024-10-15 07 48 42](https://github.com/user-attachments/assets/d3610bd3-7639-4bde-8585-4b8541cbbd7d)
![2024-10-15 07 48 36](https://github.com/user-attachments/assets/de17ff28-2376-47a1-aa4d-c2cda80eccb5)

CYD is "cheap yellow display"...just a display with an ESP-WROOM-32 mounted on the back of the [PCB](https://randomnerdtutorials.com/cheap-yellow-display-esp32-2432s028r/).
Power supply: https://www.amazon.com/gp/product/B01GEA8PQA/

**Comment by mairas** [comment]
Ah, got it. You need to provide all four N2K wires to the SH-ESP32. The SH-ESP32 N2K interface is isolated and isn't powered by the main board but by the network power.

**Comment by mairas** [comment]
And now that you mentioned, yes, I've seen that CYD device before but totally forgot. :-)

**Comment by wjquigs11** [comment]
So I can either jumper the v-link connections or move the power/GND N2K wires to the 4-pin connector, correct?

**Comment by mairas** [comment]
Move the 12V and GND wires to the 4-pin connector. If you're implementing an N2K USB gateway, the microcontroller side will be powered via USB. The tutorial is a bit old but includes correct wiring information: https://docs.hatlabs.fi/sh-esp32/pages/tutorials/nmea2000-gateway/

**Comment by wjquigs11** [comment]
I measure 12.3V across the screw terminals in the connector, but no LEDs and no voltage on any other pins (3.3V/GND).

![2024-10-15 17 01 55](https://github.com/user-attachments/assets/ac3a7dea-1533-4957-8609-caa146b9a132)

**Comment by wjquigs11** [comment]
Success! I soldered the V-link connections and it is now working.
Not sure what is the purpose of the V-link connections *not* being connected since the CAN interface does not work without that connection. Maybe a PCB wiring issue?
To clarify: 
+/- 12VDC to power connector (2-pin) does not work by itself
+/- 12VDC to CAN bus connector (4-pin) does not work by itself
Powering from USB connector enables MCU but not CAN interface
Enabling CAN interface requires 12VDC on CAN bus as well as soldered jumpers on V-link

**Comment by mairas** [comment]
TL;DR: Power the board using the USB cable and the NMEA 2000 interface using an NMEA 2000 cable. DO NOT connect the V-link pads when implementing an NMEA 2000 USB gateway!

The SH-ESP32 has two galvanically isolated sections (three if you count the optocoupler I/O, but that can be ignored for now).

1. The "main" section that includes the microcontroller, USB interface, GPIOs etc. and that is powered by either:
    1.  the 12V input or
    2.  the USB cable.
2. The CAN interface has an isolated output section that needs its own power and ground connections.

The NMEA 2000 standard mandates that NMEA 2000 devices must maintain galvanic isolation, meaning that it won't share GND, either directly or indirectly, between the NMEA 2000 network and other electrical systems.

In the case of an NMEA 2000 USB gateway, the isolation requirement means that the computer USB interface and the NMEA 2000 network must remain separated. That separation is achieved by implementing isolation in the CAN interface. Going back to the power requirements, that means that you need to power the main section using a USB cable and the CAN interface using a fully wired NMEA 2000 connector. Since those particular connections are by definition required to use an NMEA 2000 USB gateway, that is normally not an issue.

If you bridge the V-link connections, you disable galvanic isolation between the sections. Yes, your NMEA 2000 USB gateway will be powered on without the USB cable, but that gives you no benefit whatsoever. The instant you plug in a computer and plug a grounded mains plug into your inverter, you have a ground loop.

Ground loops might not have any impact, or they might cause some random dropped frames on the N2K bus, or they could interrupt communications when some other device starts malfunctioning and emitting strong interference. Since the NMEA 2000 network is often a health and safety critical system on the boat, I really don't want people to take chances.

**Comment by wjquigs11** [comment]
When I connect (powered) NMEA2000 to the 4-pin connector (no USB), the board does not turn on. No N2K LEDs, no power LED near USB connector.
If I connect powered N2K to the 4-pin connector and USB, the board powers on but I still don't see anything on N2K.
If I bridge the v-link connectors, board is powered and I see N2K traffic regardless of whether USB is connected.

**Comment by mairas** [comment]
Check the pinout of the 4-pin NMEA 2000 connector. Also, if you first connect the USB and then hot-plug NMEA 2000, you might need to press the reset button for the N2K interface to be activated. The V-link pads are not needed and should not be used in an N2k USB gateway.

**Comment by wjquigs11** [comment]
On the boat, I will not be using the USB connection, so it's OK that the V-link has to be jumpered in either case. Thanks for your help!

**Comment by wjquigs11** [comment]
Closing as resolved.
